lib/codecs/src/decoding/format/vrl.rs

```diff
@@ -0,0 +1,257 @@
+use smallvec::{smallvec, SmallVec};
+use vector_config_macros::configurable_component;
+use vector_core::config::{DataType, LogNamespace};
+use vector_core::event::{Event, TargetEvents, VrlTarget};
+use vector_core::{compile_vrl, schema};
+use vrl::compiler::state::ExternalEnv;
+use vrl::compiler::{runtime::Runtime, CompileConfig, Program, TimeZone, TypeState};
+use vrl::diagnostic::Formatter;
+use vrl::value::Kind;
+
+/// Config used to build a `VrlDeserializer`.
+#[configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct VrlDeserializerConfig {
+    /// VRL-specific decoding options.
+    vrl: VrlDeserializerOptions,
+}
+
+/// VRL-specific decoding options.
+#[configurable_component]
+#[derive(Debug, Clone, PartialEq, Eq, Derivative)]
+#[derivative(Default)]
+pub struct VrlDeserializerOptions {
+    /// The [Vector Remap Language][vrl] (VRL) program to execute for each event.
+    ///
+    /// [vrl]: https://vector.dev/docs/reference/vrl
+    pub source: String,
+
+    /// The name of the timezone to apply to timestamp conversions that do not contain an explicit
+    /// time zone. The time zone name may be any name in the [TZ database][tz_database], or `local`
+    /// to indicate system local time.
+    ///
+    /// If not set, `local` will be used.
+    ///
+    /// [tz_database]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
+    #[serde(default)]
+    #[configurable(metadata(docs::advanced))]
+    pub timezone: Option<TimeZone>,
+}
+
+impl VrlDeserializerConfig {
+    /// Build the `VrlDeserializer` from this configuration.
+    pub fn build(&self) -> vector_common::Result<VrlDeserializer> {
+        let state = TypeState {
+            local: Default::default(),
+            external: ExternalEnv::default(),
+        };
+
+        match compile_vrl(
+            &self.vrl.source,
+            &vrl::stdlib::all(),
+            &state,
+            CompileConfig::default(),
+        ) {
+            Ok(result) => Ok(VrlDeserializer {
+                program: result.program,
+                timezone: self.vrl.timezone.unwrap_or(TimeZone::Local),
+            }),
+            Err(diagnostics) => Err(Formatter::new(&self.vrl.source, diagnostics)
+                .to_string()
+                .into()),
+        }
+    }
+
+    /// Return the type of event build by this deserializer.
+    pub fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema produced by the deserializer.
+    pub fn schema_definition(&self, log_namespace: LogNamespace) -> schema::Definition {
+        match log_namespace {
+            LogNamespace::Legacy => {
+                schema::Definition::empty_legacy_namespace().unknown_fields(Kind::any())
+            }
+            LogNamespace::Vector => {
+                schema::Definition::new_with_default_metadata(Kind::any(), [log_namespace])
+            }
+        }
+    }
+}
+
+/// Deserializer that builds `Event`s from a byte frame containing protobuf.
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:**  copy paste that needs to be updated